pennylane_lightning_kokkos/src/simulator/StateVectorKokkos.hpp

```diff
@@ -702,6 +702,187 @@ template <class Precision> class StateVectorKokkos {
         }
     }
 
+    auto getIndicesAfterExclusion(const std::vector<size_t> &indicesToExclude,
+                                  size_t num_qubits) -> std::vector<size_t> {
+        std::vector<size_t> indices;
+        for (size_t i = 0; i < num_qubits; i++) {
+            indices.emplace_back(i);
+        }
+
+        for (size_t j = 0; j < indicesToExclude.size(); j++) {
+
+            const size_t excludedIndex = indicesToExclude[j];
+
+            for (size_t i = 0; i < indices.size(); i++) {
+                if (excludedIndex == indices[i])
+                    indices.erase(indices.begin() + i);
+            }
+        }
+        return indices;
+    }
+
+    auto generateBitPatterns(const std::vector<size_t> &qubitIndices,
+                             size_t num_qubits) -> std::vector<size_t> {
+
+        std::vector<size_t> indices;
+        indices.reserve(Util::exp2(qubitIndices.size()));
+        indices.emplace_back(0);
+
+        for (size_t index_it0 = 0; index_it0 < qubitIndices.size();
+             index_it0++) {
+            size_t index_it = qubitIndices.size() - 1 - index_it0;
+            const size_t value =
+                Util::maxDecimalForQubit(qubitIndices[index_it], num_qubits);
+
+            const size_t currentSize = indices.size();
+            for (size_t j = 0; j < currentSize; j++) {
+                indices.emplace_back(indices[j] + value);
+            }
+        }
+        return indices;
+    }
+
+    /**
+     * @brief Probabilities for a subset of the full system.
+     *
+     * @param wires Wires will restrict probabilities to a subset
+     * of the full system.
+     * @return Floating point std::vector with probabilities.
+     * The basis columns are rearranged according to wires.
+     */
+    std::vector<Precision> probs(const std::vector<size_t> &wires) {
+        // Determining index that would sort the vector.
+        // This information is needed later.
+        auto sorted_ind_wires = Util::sorting_indices(wires);
```

Comment:
In general the wires are sorted I think  (I'm not sure why they would be unsorted in the first place tbh), so can we have an O(N) check that they are sorted instead of NlogN sort (e.g. for loop through and see if each pair is ordered) and then put all this sorting stuff into an if statement so it doesn't run when the wires are sorted. I think this will be a very good optimization.

Reply:
Agree, it's will be O(N) at the best scenario.

Reply:
yeah that's a good point, let's just sort and check wires != sorted.wires as you say instead of the for loop